Source/DafnyCore/Verifier/ProofObligationDescription.cs

```diff
@@ -308,6 +308,45 @@ public class AssertStatement : ProofObligationDescription {
   }
 }
 
+// The Boogie version does not support custom error messages yet
+public class RequiresDescription : ProofObligationDescription {
+  private readonly string customErrMsg;
+  public override string SuccessDescription =>
+    customErrMsg is null
+      ? "the precondition always holds"
+      : $"error is impossible: {customErrMsg}";
+
+  public override string FailureDescription =>
+    customErrMsg ?? "a precondition could not be proven";
+
+  public override string ShortDescription => "requires";
+
+  public RequiresDescription([CanBeNull] string customErrMsg = null) {
+    this.customErrMsg = customErrMsg;
+  }
+}
+
+// The Boogie version does not support custom error messages yet
+public class EnsuresDescription : ProofObligationDescription {
+  private readonly string customErrMsg;
+  public override string SuccessDescription =>
+    customErrMsg is null
+      ? "this postcondition holds"
+      : $"error is impossible: {customErrMsg}";
+
+  public override string FailureDescription =>
+    customErrMsg ?? "this postcondition might not hold on a return path";
+
+  public string FailureAtPathDescription =>
+    customErrMsg ?? "a postcondition could not be proven on this return path";
```

Comment:
Isn't this second line similar to `PostconditionDescription.FailureDescription` ?

Reply:
> Isn't this second line similar to PostconditionDescription.FailureDescription ?

Yes, indeed, I added a redirection instead.